### DISCUSSION TITLE
Mobile performance (Sponza scene in AtomSampleViewer)

### DISCUSSION THREAD

**Discussion by moudgils** [o3de/o3de#2487] [General]
AtomSampleViewer - **Sponza** (RPI/SponzaBenchmark) running on **Iphone11** with LowRenderPipeline + non-msaa. Framerate is **13fps**. GPU bound. Resolution - 1080p. Vertices = 4.7million. Draw/Dispatch calls = ~170. Summary on why some passes take a long time. Only showing passes that are close to or above 1ms.

-DepthMSAAPass = 1ms

-DirectionalLightShadowmapPass0 = 1.2ms

-DirectionalLightShadowmapPass1 = 1ms

-DirectionalLightShadowmapPass2 = 0.8ms

-DirectionalLightShadowmapPass3 = 0.8ms

-EsmShadowmapsPassDirectional.Exponentiation = 2ms

-EsmShadowmapsPassDirectioonal.HorizontalGaussianFilter = 9ms

-EsmShadowmapsPassDirectioonal.VerticalGaussianFilter = 8.7ms

          - Kernel Occupancy is very high so thats good
          - For both the passes above the GPU  is spending a lot of time executing texture sampling operations and processing arithmetic instructions
          - Look into the resolution of the textures. Unable to see this info on Xcode at the moment.

-LightCullingTilePreparePass = 7.6ms

          - Kernel Occupancy is very high so thats good
          - It is spending a lot of time reading threadgroup memory
         - Recommendation -> Try to reduce the number of atomic reads?
                    -> Try to align threadgroup memory allocations to 16-byte boundary
                    -> Try to reorder your memory access patterns so that neighboring threads in a quad group write to neighboring elements in threadgroup memory.
                    -> Try to remove accesses to the same memory location from multiple threads in the same threadgroup.

-LightCullingPass = 1.85ms

         -Kernel occupancy really low.
         -ALU operations are limiting here. Register pressure?

-ForwardPass = 32ms. OUCH

        - Extremely low Pixel shader Occupancy
        - Its hard for me to tell how many waves are launched for this pass across all cores but Xcode says the shader uses 128 temporary registers and 105 Uniform registers which sounds insanely high
         - We are probably launching super low wavefronts across all cores and hence we cant hide latency from sampling textures and writing to RTs.
         - Need to seriously reduce register pressure

-SkyBoxPass = 1.43ms

-LoookModificationComposite = 1.1ms

-DisplayMapperPass.AcesOoutputTransform = 3ms

        - High fragment shader occupancy
        - Mostly all ALU operations here. Nothing obvious here 

![image](https://user-images.githubusercontent.com/47460854/127260336-a5ad7690-b651-4c43-8580-1dc2a499d478.png)

**Plan for perf work**

- Add framework to scale Atom automatically. This will enable us to switch to correct rendering pipeline and setup render settings based on hw

- Investigate Forward pass to ensure we are using the most optimal shader variants and not the root variant. 
- Do a Pass over the standard PBR to look for cheaper approximations if possible
- Investigate possibility of switching to Minimal PBR for certain meshes to help with forward pass
- If MSAA is enabled ensure that resolve happens in the forward pass itself and not done as a separate pass as this will incur a huge bandwidth related hit. 
- Disable soft shadows or look for a cheaper solution
- Lower the number of taps for Gaussian blur for mobile

**Comment by thefranke** [comment]


**Reply by rgba16f** [reply]
This was tested on an iphone 11

**Comment by CGorz** [comment]
We create an single scene with one person model, skybox & pointlight. The vertex count is half million, and lowendpipeline is used for mobile rendering. Then, we export the game to Android device (Qualcomm Snapdragon 778 processor). The final framerate is about 12fps (11fps ~ 13fps), The most time-consuming passes are listed as follows:
skybox：10ms
lightCullingpass: 6.4ms
lightCullingTilePreparePass: 6ms
MSAAResolveDepthPass: 4.7ms
AcesOutputTransform: 4.38ms
LookModificationComposite: 2.4ms
...
These performance are tested in debug mode with loose asset mode. We try to report the performance for release version (pak asset mode) . But, there is no official guildline for how to generate release apk with pak mode. So, maybe it will charge us much time to overcome it.
